### ISSUE TITLE
Add Open Telemetry support

### ISSUE THREAD

**Issue by chasers** [supabase/supavisor#93]
- should be able to send spans to my own otel vendor
- start with just a few easy otel events and we can expand from there in another issue

**Comment by abc3** [comment]
Why do we need support OTEL, given that we already provide metrics in the Prometheus format?

**Comment by J0** [comment]
Not familiar with surrounding context but based on [this article](https://www.timescale.com/blog/prometheus-vs-opentelemetry-metrics-a-complete-guide/) it sounds like Prometheus metrics is a subset of OpenTelemetry Metrics so if I had to guess maybe some customers were looking for features which were in OTEL but not PromEx

Maybe we can check with @chasers once he is back, guessing he might have more context

**Comment by chasers** [comment]
With lots of tenants we can only have so many (very few) metrics before the endpoint Prometheus scrapes gets HUGE.

e.g. the Realtime endpoint is like 10MB at the moment

OTEL gets us much more visibility into everything if instrumented well and we can get full traces of connects, and potentially everything happening in the wire protocol (not required).

**Comment by abc3** [comment]
gotcha 👍

**Comment by J0** [comment]
Makes sense, will touch up on the PR over the weekend and get back

**Comment by abc3** [comment]
@J0 The only thing is, could you make this optional? For example, use dependencies and only start using it when some environment variable, like `ENABLE_OTEL=true`, is passed

**Comment by J0** [comment]
@abc3 Sure thing, will make it optional. Haven't figured out the specifics of how to do it but was thinking maybe we can define an [OTEL build target](https://elixirforum.com/t/conditional-dependencies-in-mix-exs/48468/11) and then import only if` Mix.target() == :otel` or something.

Let me know if y'all have any preferences or suggestions

**Comment by abc3** [comment]
@J0 don't have any preferences, up to you

**Comment by J0** [comment]
Hey

@abc3 am a little stuck on the conditional imports - seems like Dialyzer will complain if the imports are conditional since the [ Application `start` will be relying on dependencies that won't be bundled in when `Mix.Target()` is not set to OpenTelemetry](https://github.com/supabase/supavisor/pull/102/files#diff-05b9775ba252856439da7e8258eb1d1ec07237573d0860ddb8593075a65fb5d7R13)
Wondering if y'all have any suggestions there with respect to conditional imports.


Alternatively, is there any chance we can keep the Opentelemetry dependencies but continue to only start using it when some environment variable, like ENABLE_OTEL=true, is passed?

**Comment by abc3** [comment]
> Wondering if y'all have any suggestions there with respect to conditional imports.

you can use macros or ignore warnings through @dialyzer module attribute

> Alternatively, is there any chance we can keep the Opentelemetry dependencies but continue to only start using it when some environment variable, like ENABLE_OTEL=true, is passed?

I would prefer to avoid that approach

**Comment by J0** [comment]
Gotcha, thanks! Will try that in the evening or early tmrw

**Comment by J0** [comment]
Hey, 

Wanted to circle back on this - I wasn't able to figure out how to properly make the imports conditional. If anyone wants to jump on this please feel free - otherwise will look again in the future whenever a slot frees up

**Comment by janpio** [comment]
Original issue descriptions seems to be about spans (so Otel tracing), while discussion then pivoted to metrics. Was that intentional?

Otel spans / telemetry would be very interesting to be able to add deep understanding to each query that is being handled.

**Comment by chasers** [comment]
@janpio no, this is still re: implementing Otel tracing along the query pipeline. 

We have already implemented a lot of metrics which are [exposed](https://github.com/supabase/supavisor/blob/4265b5143a4577f064b5fb1d7b23c6cd6bc36c18/lib/supavisor_web/router.ex#L51) at `/metrics` (for the whole system) and `/metrics/:external_id` (filtered per tenant).

**Comment by dkuku** [comment]
As this is a topic about opentelemetry I will ask here:
Did you consider adding [sqlcommenter](https://google.github.io/sqlcommenter/) support - [this is a part of opentelemetry](https://cloud.google.com/blog/products/databases/sqlcommenter-merges-with-opentelemetry).

**Comment by ConProgramming** [comment]
Think this would be huge. An offering like [highlight.io](https://highlight.io) inside Supabase would be incredible.

**Comment by mikemiller-avoca** [comment]
This would be immensely useful. It would allow us to track down deadlocks to the exact line of application code that caused them.
